lib/ruby_lsp/requests/hover.rb

```diff
@@ -0,0 +1,72 @@
+# typed: strict
+# frozen_string_literal: true
+
+require "ruby_lsp/requests/support/rails_document_client"
+
+module RubyLsp
+  module Requests
+    # ![Hover demo](../../misc/rails_document_link_hover.gif)
+    #
+    # The [hover request](https://microsoft.github.io/language-server-protocol/specification#textDocument_hover)
+    # renders a clickable link to the code's official documentation.
+    # It currently only supports Rails' documentation: when hovering over Rails DSLs/constants under certain paths,
+    # like `before_save :callback` in `models/post.rb`, it generates a link to `before_save`'s API documentation.
+    #
+    # # Example
+    #
+    # ```ruby
+    # class Post < ApplicationRecord
+    #   before_save :do_something # when hovering on before_save, the link will be rendered
+    # end
+    # ```
+    class Hover < BaseRequest
+      extend T::Sig
+
+      sig { params(document: Document, position: Document::PositionShape).void }
+      def initialize(document, position)
+        super(document)
+
+        @position = T.let(Document::Scanner.new(document.source).find_position(position), Integer)
+      end
+
+      sig { override.returns(T.nilable(LanguageServer::Protocol::Interface::Hover)) }
+      def run
+        target, _ = locate_node_and_parent(
+          T.must(@document.tree), [SyntaxTree::Command, SyntaxTree::FCall, SyntaxTree::ConstPathRef], @position
+        )
+
+        case target
+        when SyntaxTree::Command
```

Comment:
The reason I had suggested checking the node locations in the visits is because it would conform to the style we're using in most requests and because the code is a bit easier to read.

Is there a reason why we can't use the visitor pattern here? What's the advantage of using this locate node strategy?

Reply:
Because I agree with what @paracycle said [here](https://github.com/Shopify/ruby-lsp/pull/313#discussion_r984056705), it's more clear to just return the result than mutate and return the `@hover` ivar.

In other requests, we go through nodes to collect data (highlights, tokens...etc.), push them to a state, and return the final state. So that pattern makes sense.

But in this case, we don't have a bunch of data to collect , nor do we have multiple nodes to check. We only care about one node, which we can locate right away. This meant having that `@hover` state becomes unnecessary. Also, based on how other requests work, seeing a visitor pattern would give me an impression that we may match multiple nodes, which it's not the case. 

Therefore, I actually think the current implementation is easier to read and understand.